### ISSUE TITLE
The Loss Increase Strangely

### ISSUE THREAD

**Issue by qwenqw** [GeorgeSeif/Semantic-Segmentation-Suite#123]
------------------------

## Please specify the following information when submitting an issue:
- **What are your command line arguments?:**python train.py --dataset Myown --crop_height 960  --crop_width 960  --model PSPNet --frontend ResNet50
- **Have you written any custom code?**: I adjusted the dataset to my own data. And I changed the file class_dict.csv. The other parameters as default. (lr=0.0001, decay=0.995)
- **What have you done to try and solve this issue?:**No idea. I am not sure what is wrong.
- **TensorFlow version?**:1.2.0  CUDA=8.0 Cudnn=5.1

### Describe the problem

Hi, I just trained on my own dataset with 130 epoch. The training didn't fail. But during the training, the average loss kept increasing. And the validation IoU is very unstable. The result as follows.

Moreover, my own dataset has 6 classes, including the background. And the training set has 880 images, the testing set has 200 images.
Now I am really confused. Am I doing something wrong?
Please help me solve it. 
Thank you all in advance.

![image](https://user-images.githubusercontent.com/23628614/47135656-bde34f00-d2e3-11e8-8f33-8daa51e78cab.png)
![image](https://user-images.githubusercontent.com/23628614/47135678-d05d8880-d2e3-11e8-9547-6338438f577a.png)
![image](https://user-images.githubusercontent.com/23628614/47135733-10247000-d2e4-11e8-820e-441e76dcf54a.png)

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

[2018-10-17 18:27:09] Remaining training time = 26 hours 24 minutes 29 seconds

[2018-10-17 18:27:22] Epoch = 130 Count = 20 Current_Loss = 0.0786 Time = 11.95
[2018-10-17 18:27:34] Epoch = 130 Count = 40 Current_Loss = 0.0495 Time = 11.95
[2018-10-17 18:27:45] Epoch = 130 Count = 60 Current_Loss = 16.9121 Time = 11.98
[2018-10-17 18:27:57] Epoch = 130 Count = 80 Current_Loss = 0.3031 Time = 11.95
[2018-10-17 18:28:09] Epoch = 130 Count = 100 Current_Loss = 0.0254 Time = 11.97
[2018-10-17 18:28:21] Epoch = 130 Count = 120 Current_Loss = 13.6251 Time = 11.93
[2018-10-17 18:28:33] Epoch = 130 Count = 140 Current_Loss = 2.3520 Time = 11.96
[2018-10-17 18:28:45] Epoch = 130 Count = 160 Current_Loss = 0.0803 Time = 11.96
[2018-10-17 18:28:57] Epoch = 130 Count = 180 Current_Loss = 0.0197 Time = 11.95
[2018-10-17 18:29:09] Epoch = 130 Count = 200 Current_Loss = 0.0311 Time = 11.98
[2018-10-17 18:29:21] Epoch = 130 Count = 220 Current_Loss = 16.3048 Time = 11.97
[2018-10-17 18:29:33] Epoch = 130 Count = 240 Current_Loss = 102.0692 Time = 11.99
[2018-10-17 18:29:45] Epoch = 130 Count = 260 Current_Loss = 0.0175 Time = 12.01
[2018-10-17 18:29:57] Epoch = 130 Count = 280 Current_Loss = 115.6432 Time = 11.96
[2018-10-17 18:30:09] Epoch = 130 Count = 300 Current_Loss = 0.0895 Time = 12.01
[2018-10-17 18:30:21] Epoch = 130 Count = 320 Current_Loss = 0.4005 Time = 12.01
[2018-10-17 18:30:33] Epoch = 130 Count = 340 Current_Loss = 0.0322 Time = 12.00
[2018-10-17 18:30:45] Epoch = 130 Count = 360 Current_Loss = 0.6240 Time = 11.98
[2018-10-17 18:30:57] Epoch = 130 Count = 380 Current_Loss = 2.2971 Time = 11.96
[2018-10-17 18:31:09] Epoch = 130 Count = 400 Current_Loss = 4.7438 Time = 11.96
[2018-10-17 18:31:21] Epoch = 130 Count = 420 Current_Loss = 0.1266 Time = 11.99
[2018-10-17 18:31:33] Epoch = 130 Count = 440 Current_Loss = 5.0472 Time = 12.03
[2018-10-17 18:31:45] Epoch = 130 Count = 460 Current_Loss = 3.7820 Time = 11.99
[2018-10-17 18:31:57] Epoch = 130 Count = 480 Current_Loss = 0.0595 Time = 11.94
[2018-10-17 18:32:09] Epoch = 130 Count = 500 Current_Loss = 2.3438 Time = 11.99
[2018-10-17 18:32:21] Epoch = 130 Count = 520 Current_Loss = 16.4954 Time = 11.97
[2018-10-17 18:32:33] Epoch = 130 Count = 540 Current_Loss = 0.3656 Time = 11.96
[2018-10-17 18:32:45] Epoch = 130 Count = 560 Current_Loss = 1.8269 Time = 11.95
[2018-10-17 18:32:57] Epoch = 130 Count = 580 Current_Loss = 4.8849 Time = 11.97
[2018-10-17 18:33:09] Epoch = 130 Count = 600 Current_Loss = 0.2584 Time = 11.93
[2018-10-17 18:33:21] Epoch = 130 Count = 620 Current_Loss = 0.3795 Time = 11.97
[2018-10-17 18:33:33] Epoch = 130 Count = 640 Current_Loss = 0.1069 Time = 11.97
[2018-10-17 18:33:45] Epoch = 130 Count = 660 Current_Loss = 1.8307 Time = 11.98
[2018-10-17 18:33:57] Epoch = 130 Count = 680 Current_Loss = 0.0496 Time = 11.92
[2018-10-17 18:34:09] Epoch = 130 Count = 700 Current_Loss = 0.4369 Time = 11.99
[2018-10-17 18:34:21] Epoch = 130 Count = 720 Current_Loss = 0.0194 Time = 11.98
[2018-10-17 18:34:33] Epoch = 130 Count = 740 Current_Loss = 2.5540 Time = 11.99
[2018-10-17 18:34:45] Epoch = 130 Count = 760 Current_Loss = 1.3295 Time = 11.96
[2018-10-17 18:34:56] Epoch = 130 Count = 780 Current_Loss = 0.0537 Time = 11.98
[2018-10-17 18:35:08] Epoch = 130 Count = 800 Current_Loss = 27.3824 Time = 11.97
[2018-10-17 18:35:20] Epoch = 130 Count = 820 Current_Loss = 4.8048 Time = 11.95
[2018-10-17 18:35:32] Epoch = 130 Count = 840 Current_Loss = 0.0680 Time = 11.94
[2018-10-17 18:35:44] Epoch = 130 Count = 860 Current_Loss = 2.4314 Time = 12.01
[2018-10-17 18:35:56] Epoch = 130 Count = 880 Current_Loss = 10.6609 Time = 11.96
Saving latest checkpoint
Saving checkpoint for this epoch
Performing validation

**Comment by qwenqw** [comment]
By the way, how can I mannually set the class weight? in train.py or utils.py?

**Comment by elliestath** [comment]
hey, I had similar results while training the CamVid dataset with lower resolution images 192*144 pixel, keeping the batch size 1 and the learning rate 0.001. My crop height and width is 64 and I trained for 300 epochs.

**Comment by GeorgeSeif** [comment]
@qwenqw For your issue, I'd definitely try a different training set up with a more aggressive learning rate decay. As you can see from your graph, the loss was okay at first but exploded later. An exploding loss usually indicates a learning rate that is too high

@elliestath You might be having a combination of problems. The first could be the same as @qwenqw with the learning rate. But also, using such small crops tends to cause unstable learning since the crops in this code do not scan across the image. Instead, I take a single crop per image.

**Comment by elliestath** [comment]
@GeorgeSeif Thank you! What do you mean by "single crop per image"? 
I tried reducing the lr = 0.0001 also, and a crop size of 128x128, nothing changed, my loss is actually increasing every epoch. Here is a trial I made but had to stop the learning after a few epochs, since the loss was too high already.
![loss_vs_epochs](https://user-images.githubusercontent.com/11046400/47231941-9a490300-d3ce-11e8-8909-31575a3c7e84.png)

**Comment by GeorgeSeif** [comment]
By "single crop per image" I mean that if I have 1000 images, each epoch will also use only 1000 images cropped from your training data. 

Yeah judging from the graph I would increase crop size, especially with the heavy pooling done by PSPNet. You can also try another network that has less pooling!

**Comment by qwenqw** [comment]
@GeorgeSeif Hi, thank you for replying. I also tried reducing the learning rate to 0.001 or 0.0001. But the LovaszSoftmax loss was still abnormal. Is the learning rate still too high?  

After that, I changed the loss function to softmax_cross_entropy_with_logits.  I also used class weights, following  commit 43561f7. And the result seems right.

Maybe the LovaszSoftmax loss is not suitable for PSPNet ? Have you ever trained PSPNet with LovaszSoftmax loss?

**Comment by GeorgeSeif** [comment]
You didn't specify that you were using the `LovaszSoftmax` before! There's a reason I took it out of the code, it doesn't work well at all. Please stick to the `softmax_cross_entropy_with_logits`, with or without your own class balancing as that one has been tested well to work. 

Closing as this is not an issue with the current code base
